docs/architecture.md

```diff
@@ -0,0 +1,108 @@
+
+* The goal is to make it easy for the user to configure the agent
+* That's done via JSON config files (well, there is a small extension to the JSON parser so the
+  config file supports JavaScript style comments)
+* Major configuration options can also be set via environment variables (that was done to better
+  support modern Docker / Kubernetes environments which usually utilize environment variables for
+  config option)
+* It supports "standard" Linux approach of a single configuration file (
+  ``/etc/scalyr-agent-2/agent.json``) with support for merging in options from "add-on"
+  configuration files from ``/etc/scalyr-agent-2/agent.d/*.json``)
+
+#### Platform / OS Abstraction
+
+* Agent supports Linux and Windows
+* Platform specific functionality is available in ``platform_*.py`` files
+
+### 2. Copy Manager ([copy_manager.py](https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/copy_manager.py))
+
+* It's responsible for reading the log content from disk and sending it to the Scalyr API
+* It's told what it should be reading / copying.
+  * It get's a list of log files which it should be copying / uploading to Scalyr + the
+    corresponding log file settings (what parser should be used, any reduction or samplying rules,
+    arbitrary attributes associated with that log file, etc.)
+* It also performs glob expansion (user can specify a glob for files to monitor, e.g.
+  ``/var/log/mongodb/*.log``)
+* Once every 30 seconds it checks and expands globs if new files for a particular glob are found.
+  It adds those files to the files watched (glob expansion poll value can be configured via config
+  option)
+* It relies on a couple of other abstractions from log_processing.py which are documented below
+
+#### LogMatcher ([log_processing.py](https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/log_processing.py))
+
+* Responsible for resolving globs. There is an instance of ``LogMatcher`` for each glob pattern (it
+  takes whole configuration for that log file glob entry).
+* When a new file matching a glob is found, it instantiates LogProcessor class
+
+#### LogFileProcessor ([log_processing.py](https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/log_processing.py))
+
+* Responsible for copying one physical file content to the Scalyr API.
+* Reads bytes out of the log file and inserts it into the API request.
+* It also manages check points.
+* Check point represents up to which point in file we have copied bytes so far. To take log
+  rotation, file moving and other possible changes into account, it relies on the ``LogFileIterator``
+  abstraction.
+
+#### LogFileIterator ([log_processing.py](https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/log_processing.py))
+
+* It's responsible to watch a single file path and give an illusion that you are iterating over
+  all the bytes in a log file even if that log file has been rotated.
```

Comment:
Here I would say "iterating over all log lines in a log file".  I just want to make the point the abstraction iterates at the line level.